Introduce Bento sanitizer to handle with `bento`-prefixed components

westonruter opened the pull request:
## Summary

Bento components will be available not only as `amp`-prefixed components, but also with the `bento` prefix. Consider this following example which can be pasted into a Custom HTML block on a site without the AMP plugin running at all:

```html
<script
	type="module"
	async
	src="https://cdn.ampproject.org/bento.mjs"
></script>
<script nomodule src="https://cdn.ampproject.org/bento.js"></script>
<script
	type="module"
	async
	src="https://cdn.ampproject.org/v0/bento-accordion-1.0.mjs"
></script>
<script
	nomodule
	async
	src="https://cdn.ampproject.org/v0/bento-accordion-1.0.js"
></script>
<link
	rel="stylesheet"
	type="text/css"
	href="https://cdn.ampproject.org/v0/bento-accordion-1.0.css"
/>

<p>The following should have a green outline.</p>

<bento-accordion id="my-accordion">
	<section>
		<h2>Section 1</h2>
		<div>Content in section 1.</div>
	</section>
	<section>
		<h2>Section 2</h2>
		<div>Content in section 2.</div>
	</section>
	<!-- Expanded on page load due to attribute: -->
	<section expanded>
		<h2>Section 3</h2>
		<div>Content in section 3.</div>
	</section>
</bento-accordion>

<!-- Note the use of bento-accordion tag in the selector. When converted to amp-accordion, the selector must also be converted. -->
<style>
	#my-accordion {
		display: block;
		outline: solid 2px red;
	}
	bento-accordion#my-accordion {
		outline: solid 2px green;
	}
</style>
```

On a non-AMP page, this is output as expected:

![image](https://user-images.githubusercontent.com/134745/141928507-dc3f825d-d409-4f58-906b-44f6b61a4f1c.png)

However, when viewing the AMP page it completely breaks:

![image](https://user-images.githubusercontent.com/134745/141928594-29bca502-c7aa-48df-9332-3f004155c556.png)

Validating the page shows that there are many errors detected, resulting in the `bento-accordion` and Bento stylesheet and scripts being removed:

![image](https://user-images.githubusercontent.com/134745/141928745-c326cf4b-7f22-472a-92e6-afabc4c806cc.png)

At the moment, all of the Bento components are just aliases for corresponding v1.0 AMP components. Therefore, in order to seamlessly incorporate Bento components into valid AMP pages, what is needed is to rename the `bento` prefix to `amp` when there is a corresponding component available. So there is an `bento-accordion` component and a corresponding [`amp-accordion` v1.0](https://amp.dev/documentation/components/amp-accordion-v1.0/?format=websites) which is also the Bento version, as opposed to [v0.1](https://amp.dev/documentation/components/amp-accordion/?format=websites) which is the classic version. So we can just rename `bento-accordion` to `amp-accordion` which will ensure that it is recognized as valid AMP markup and thus won't be sanitized from the page. Any CSS selectors referencing `bento-accordion` are also updated to use `amp-accordion`. When such a conversion is done, any Bento-specific stylesheets and scripts can also be removed from the document since the auto-extension importing AMP will then automatically add the required markup.

So the above HTML is converted into the following on a valid AMP page:

```html
<style amp-custom>#my-accordion{display:block;outline:solid 2px red}amp-accordion#my-accordion{outline:solid 2px green}</style>
<script src="https://cdn.ampproject.org/v0/amp-accordion-1.0.mjs" async="" custom-element="amp-accordion" type="module" crossorigin="anonymous"></script>
<script async nomodule src="https://cdn.ampproject.org/v0/amp-accordion-1.0.js" crossorigin="anonymous" custom-element="amp-accordion"></script>

...

<p>The following should have a green outline.</p>

<amp-accordion id="my-accordion" class="i-amphtml-layout-container" i-amphtml-layout="container">
	<section>
		<h2>Section 1</h2>
		<div>Content in section 1.</div>
	</section>
	<section>
		<h2>Section 2</h2>
		<div>Content in section 2.</div>
	</section>
	<!-- Expanded on page load due to attribute: -->
	<section expanded>
		<h2>Section 3</h2>
		<div>Content in section 3.</div>
	</section>
</amp-accordion>

<!-- Note the use of bento-accordion tag in the selector. When converted to amp-accordion, the selector must also be converted. -->
```

This PR also accounts for future Bento components which do not have AMP aliases. When such a component is encountered, the component along with its dependent scripts and stylesheet will not marked with `data-px-verified` so that the plugin's sanitizer will not later strip it out. Such pages with non-AMP Bento components will not be be marked as `amp` pages as they are not valid AMP, although they will still incorporate the AMP runtime and will go through the AMP Optimizer. This is the “Moderate” sandboxing level introduced in #6443. 

### Fix handling of ID attributes during sanitizer conversion

In 263501f3e7945ad27967d5a23c466cf118a023cf I worked around [Unexpected Handling of Element IDs in PHP DOM](https://weston.ruter.net/2021/11/19/unexpected-handling-of-element-ids-in-php-dom/). In short, it turns out that PHP DOM does not handle element IDs the same way as browser DOM does. Therefore, in order to element conversions successfully, it's important to _remove_ the `id` attribute from the old element before adding it to the new element.

### Fix JS error on Validated URL screen when there are no Validation Errors

I discovered on the Validated URL screen that the rows wouldn't expand in the Stylesheets metabox. Looking at the console I saw an error coming from:

https://github.com/ampproject/amp-wp/blob/5405daa38e65f0ec16ffc920014d0110b03ee773/assets/src/amp-validation/amp-validated-url-post-edit-screen.js#L278

This is for this search box:

![image](https://user-images.githubusercontent.com/134745/143078420-a20b2784-2ef5-4ae1-8f90-b6dedfe890ce.png)

The issue is that when there are no Validation Errors, then no search box is added to the page:

![image](https://user-images.githubusercontent.com/134745/143078492-f9cbf200-acae-409c-9fc3-1aa27d7b072b.png)

And this causes an error. Fixed in e13cc99bc2764aff2021ef2eb4a8405bb366c1f3.

## Checklist

- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

westonruter committed:
Introduce Bento sanitizer to deal with BentoJS components

westonruter commented:
- [x] The CSS tree-shaker is actually removing `#my-accordion{display:block;outline:solid 2px red}` when it shouldn't be.

westonruter commented:
- [x] Ensure compatibility with Gutenberg Bento. See https://github.com/swissspidy/gutenberg-bento/issues/33.

westonruter committed:
Fix handling of ID attributes during sanitizer conversion

westonruter committed:
Prevent JS error on validated URL screen when no validation errors present

westonruter committed:
Prevent bento components from being treated as light shadow DOM wrappers

westonruter committed:
Improve Bento sanitizer and add tests

* Fix exempting non-AMP Bento stylesheet
* Ensure scripts are removed if no Bento components present
* Prefer bento when any Bento component is discovered (not just when conversion happened)

westonruter committed:
Improve test coverage for ID attribute changes

westonruter committed:
Further improve test converage and remove dead code

westonruter committed:
Add support for converting Bento component dimensions from inline styles to AMP layout

delawski commented on includes/sanitizers/class-amp-base-sanitizer.php:
```diff
@@ -128,12 +128,46 @@ public static function add_buffering_hooks( $args = [] ) {} // phpcs:ignore Vari
+	 * Sometimes AMP components serve as wrappers for native elements, like `amp-img` for `img`. In this case, when this
+	 * is the case, authors sometimes will want to style the shadow element (such as to set object-fit). Normally if a
+	 * selector contains `img` then the style sanitizer will always convert this to `amp-img` (and `amp-anim`), which
+	 * may break the author's intended selector target. So the when using a sanitizer's selector conversion mapping to
```
```suggestion
	 * may break the author's intended selector target. So when using a sanitizer's selector conversion mapping to
```

delawski commented on includes/sanitizers/class-amp-base-sanitizer.php:
```diff
@@ -128,12 +128,46 @@ public static function add_buffering_hooks( $args = [] ) {} // phpcs:ignore Vari
+	/**
+	 * Determine whether the resulting AMP element uses a "light" shadow DOM.
+	 *
+	 * Sometimes AMP components serve as wrappers for native elements, like `amp-img` for `img`. In this case, when this
```
```suggestion
	 * Sometimes AMP components serve as wrappers for native elements, like `amp-img` for `img`. When this
```

delawski commented on includes/sanitizers/class-amp-bento-sanitizer.php:
```diff
@@ -0,0 +1,264 @@
+		foreach ( $scripts as $script ) {
+			/** @var Element $script */
+			$basename = basename( wp_parse_url( $script->getAttribute( Attribute::SRC ), PHP_URL_PATH ) );
+			if ( ! preg_match( '#^(bento-.*?)-\d+\.\d+\.m?js#', $basename, $matches ) ) {
```
Is there a reason why the script's `src` is parsed in a different way than the link's `href` attribute above, i.e.:
```php
$bento_name = preg_replace( '/-\d+\.\d+\.m?js$/', '', $basename );
```

westonruter replied:
Humm, I don't think there is a reason actually.

westonruter replied:
I think the former should actually use this approach instead, as then it it wouldn't accidentally set `$bento_name` to be something completely erroneous.

westonruter replied:
See 25c4e313d313be46695c2cc37597407b20fd1d41

delawski commented on includes/sanitizers/class-amp-bento-sanitizer.php:
```diff
@@ -0,0 +1,264 @@
+			$layout_attributes[ Attribute::LAYOUT ] = Layout::FIXED_HEIGHT;
+			unset( $styles['width'] );
+		} else {
+			$height = new CssLength( $styles['width'] );
+			$height->validate( false, false );
+			if ( $height->isValid() ) {
+				$layout_attributes[ Attribute::WIDTH ] = $height->getNumeral() . ( $height->getUnit() !== 'px' ? $height->getUnit() : '' );
+				unset( $styles['width'] );
+			}
```
Should the variable be named `$width` instead of `$height`?

westonruter replied:
Ooof, yes, I think you're right. Copypasta my bad.

```suggestion
			$width = new CssLength( $styles['width'] );
			$width->validate( false, false );
			if ( $width->isValid() ) {
				$layout_attributes[ Attribute::WIDTH ] = $width->getNumeral() . ( $width->getUnit() !== 'px' ? $width->getUnit() : '' );
				unset( $styles['width'] );
			}
```

delawski reviewed (commented):
I went through the code and left a few minor comments.

I didn't get a chance to test it out locally, but will do so on Monday.

westonruter committed:
Fix typos in comments

Co-authored-by: Piotr Delawski <delawski@users.noreply.github.com>

westonruter committed:
Fix variable name

westonruter committed:
Harmonize Bento component name parsing

delawski commented:
I've tested the code locally. The provided sample code displaying an accordion works as expected.

I also tried adding some [`bento-mathml`](https://bentojs.dev/components/bento-mathml/) formulas. I used a code from the documentation (inside a Custom HTML block):

```html
<script type="module" src="https://cdn.ampproject.org/bento.mjs" crossorigin="anonymous"></script>
<script nomodule src="https://cdn.ampproject.org/bento.js" crossorigin="anonymous"></script>
<script type="module" src="https://cdn.ampproject.org/v0/bento-mathml-1.0.mjs" crossorigin="anonymous"></script>
<script nomodule src="https://cdn.ampproject.org/v0/bento-mathml-1.0.js" crossorigin="anonymous"></script>
<link rel="stylesheet" href="https://cdn.ampproject.org/v0/bento-mathml-1.0.css" crossorigin="anonymous">

<h2>The Quadratic Formula</h2>
<bento-mathml
  style="height: 40px"
  data-formula="\[x = {-b \pm \sqrt{b^2-4ac} \over 2a}.\]"
></bento-mathml>

<h2>Inline formula</h2>
<p>
  This is an example of a formula,
  <bento-mathml
    style="height: 40px; width: 147px"
    inline
    data-formula="\[x = {-b \pm \sqrt{b^2-4ac} \over 2a}.\]"
  ></bento-mathml>
  placed inline in the middle of a block of text.
</p>
```

The AMP version doesn't render the formulas.

| Frontend | AMP Validated URL |
| --- | --- | 
| ![Screenshot 2021-11-29 at 15 19 55](https://user-images.githubusercontent.com/478735/143884382-ef011bfc-4c97-465f-9d77-c6e8b9297123.png) | ![Screenshot 2021-11-29 at 15 20 35](https://user-images.githubusercontent.com/478735/143884511-5274d30f-359f-4b70-a21c-0ba4bcd91d14.png) |

When I removed the inline `style` attributes, and added CSS in a separate `<style>` element, I got no AMP error and the formulas were rendered as expected.

```html
<style>
  bento-mathml {
    width: 150px;
    height: 40px;
  }
</style>
```
![Screenshot 2021-11-29 at 15 52 34](https://user-images.githubusercontent.com/478735/143889824-b9b1b900-6cde-4d04-a3ab-e8e32b31e409.png)

Is that expected behaviour?

westonruter commented:
> Is that expected behaviour?

No, that's not expected. The problem appears to be that the sanitizer is adding layouts that the component doesn't support:

![image](https://user-images.githubusercontent.com/134745/143910726-75f5bf6a-151b-417b-af2d-dfcf3fa40b16.png)

I need to make sure that is factored into the sanitizer logic.

westonruter commented:
> When I removed the inline `style` attributes, and added CSS in a separate `<style>` element, I got no AMP error and the formulas were rendered as expected.

This is because the default layout then becomes `container`, but since the element has no initial content prior to rendering the formula, then it gets no height. But seems the `amp-mathml` component _only_ supports the `container` layout so it should never have `width`/`height` attributes added to it.

Working on a fix.

westonruter committed:
Update AMP_Base_Sanitier::set_layout() to account for 100% width/height styles

westonruter committed:
Validate supported layouts for converted Bento components; add support for fill and nodisplay

westonruter committed:
Derive intrinsic layout when max-width defined

westonruter commented:
OK, this is now much more robust.

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into add/bento-js-support

* 'develop' of github.com:ampproject/amp-wp:
  Use CF7 instead of Autoptimize for AMP-incompatible plugin in E2E tests
  Replace autoptimize with akismet for E2E tests
  Pin NodeJS to v14 to match WP core and Gutenberg
  Regenerate package-lock.json at lockfileVersion 1
  Update Gutenberg package dependencies
  Update Gutenberg package dependencies

delawski commented on includes/sanitizers/class-amp-bento-sanitizer.php:
```diff
@@ -229,53 +230,144 @@ private function get_bento_component_name_from_url( $url ) {
+		}
+
+		// If nodisplay is the only supported layout or nodisplay is supported the
+		// element is hidden, then give it the nodisplay layout.
```
```suggestion
		// If nodisplay is the only supported layout or nodisplay is supported and
		// the element is hidden, then give it the nodisplay layout.
```

delawski commented on includes/sanitizers/class-amp-bento-sanitizer.php:
```diff
@@ -229,53 +230,144 @@ private function get_bento_component_name_from_url( $url ) {
+			(
+				$amp_element->hasAttribute( Attribute::HIDDEN )
+				&&
+				in_array( Layout::TO_SPEC[ Layout::NODISPLAY ], $supported_layouts )
```
Do we want it to be a strict check?

```suggestion
				in_array( Layout::TO_SPEC[ Layout::NODISPLAY ], $supported_layouts, true )
```

westonruter replied:
Yes, good catch.

delawski reviewed (commented):
With the latest changes, `bento-mathml` displays correctly when there is no inline `style` attribute.

I've also tested `bento-instagram` both with explicit dimensions provided in an inline `style` attribute and with `position: absolute; width: 100%; height: 100%` (in order to trigger a `layout="fill"`). In both cases the Instagram embed rendered as expected.

westonruter committed:
Use strict check for in_array() and improve comment

Co-authored-by: Piotr Delawski <delawski@users.noreply.github.com>

milindmore22 commented:
All looks good, I tested few random components date-countdown, base carousel, twitter, WordPress Embed All looks good except base carousel 
When I tried example for [bento Carousel from the docs](https://bentojs.dev/components/bento-carousel/) which has no layout or inline style defined it didn't show until I add layout.

here is what I tried 
```html
<script type="module" src="https://cdn.ampproject.org/bento.mjs" crossorigin="anonymous"></script>
<script nomodule="" src="https://cdn.ampproject.org/bento.js" crossorigin="anonymous"></script>

<script type="module" src="https://cdn.ampproject.org/v0/bento-base-carousel-1.0.mjs" crossorigin="anonymous"></script>
<script nomodule="" src="https://cdn.ampproject.org/v0/bento-base-carousel-1.0.js" crossorigin="anonymous"></script>

<link rel="stylesheet" href="https://cdn.ampproject.org/v0/bento-base-carousel-1.0.css" crossorigin="anonymous">
  <bento-base-carousel id="my-carousel">
    <div class="red"></div>
    <div class="blue"></div>
    <div class="green"></div>
  </bento-base-carousel>
```
CSS for it you can add it Customizer
```css
bento-base-carousel,
    bento-base-carousel > div {
      aspect-ratio: 4/1;
}
.red {
      background: darkred;
}
.blue {
      background: steelblue;
}
.green {
      background: seagreen;
}
```
This won't show unless you add layout for it, it does support all layouts

westonruter commented:
> has no layout or inline style defined it didn't show until I add layout

@milindmore22 Yes, this is somewhat of a known limitation. In order for that `bento-base-carousel` example to automatically convert to `amp-base-carousel`, the `aspect-ratio` has to be added to as a `style` attribute on the element itself. Otherwise the sanitizer can't determine the layout to obtain. This is shown in the tests:

https://github.com/ampproject/amp-wp/blob/1d435a2d9314d7915d0d1182a09fe98badc4c08d/tests/php/test-amp-bento-sanitizer.php#L223

is converted to:

https://github.com/ampproject/amp-wp/blob/1d435a2d9314d7915d0d1182a09fe98badc4c08d/tests/php/test-amp-bento-sanitizer.php#L255

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into add/bento-js-support

* 'develop' of github.com:ampproject/amp-wp: (45 commits)
  Update Gutenberg package dependencies
  Update CODEOWNERS
  Update amphtml spec to 2111242025000
  Stop mapping E2E test plugins dir to container volume
  Add demo plugin ZIP archive
  Fix script command
  Fix lint issues
  Ensure React state updates are not performed on unmounted components
  Use different Jest configs for E2E tests; use `.env` config file
  Refactor "Other Settings" E2E tests
  Ensure selector is available before scrolling element into view
  Use more robust internal activate/deactivate functions in E2E tests
  Introduce general `saveSettings` util function for E2E tests
  Remove unnecessary cleanup functions in E2E tests
  Install Hestia theme globally as it is used in multiple tests
  Use local demo plugin in E2E tests
  Ensure disabled back button is covered by E2E tests
  Add `SiteScanContextProvider` mock to prevent unit test errors
  Ensure Site Scan progress bar never goes backwards
  Prevent going back in Wizard if Site Scan is in progress
  ...

westonruter committed:
Fix phpcs alignment warning

westonruter committed:
Add test case for Bento lacking layout attributes; improve test coverage

milindmore22 commented:
**QA Passed**

Everything looks good 👍🏼 

✅   Bento components are renamed to the AMP components.
✅   AMP 1.0 scripts are included for the bento components.
✅   CSS selector with bento alias are converted to amp alias.
